### PR TITLE
Add Advanced property and put BeforeSeal function in it

### PR DESCRIPTION
**Pull request by TylerCarlson1**
As discussed in #1866

### CODE REVIEW

**Comment by lbargaoanu** [comment]
A test would help.

**Comment by lbargaoanu** [comment]
Maybe it's just me, but I don't understand the test :) Maybe it's even correct, but I cannot tell by reading the code.

**Comment by TylerCarlson1** [comment]
Oh yea you are right I didn't even notice that.  By default _sealed is false and doesn't really prove the method was called.
I can make _sealed nullable, so you know it got set.

**Comment by lbargaoanu** [comment]
It makes sense.

File: src/AutoMapper/AdvancedConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using System;
+
+namespace AutoMapper
+{
+    public class AdvancedConfiguration
+    {
+        /// <summary>
+        /// Action called against the IConfigurationProvider before it gets sealed
+        /// </summary>
+        public Action<IConfigurationProvider> BeforeSeal { get; set; }
```
Comment (TylerCarlson1) [inline src/AutoMapper/AdvancedConfiguration.cs]: Yea I thought about that.  But I got the thing to work with AM.Collections with a single action, so haven't changed it yet.

File: src/AutoMapper/MapperConfiguration.cs [hunk 2]
```diff
@@ -37,7 +37,8 @@ public MapperConfiguration(MapperConfigurationExpression configurationExpression
             Configuration = new ProfileMap(configurationExpression);
             Profiles = new[] { Configuration }.Concat(configurationExpression.Profiles.Select(p => new ProfileMap(p, configurationExpression))).ToArray();
 
-            configurationExpression.Advanced.BeforeSeal?.Invoke(this);
+            foreach (var beforeSealAction in configurationExpression.Advanced.BeforeSealActions)
+                beforeSealAction?.Invoke(this);
```
Comment (lbargaoanu) [inline src/AutoMapper/MapperConfiguration.cs]: I think this code should be inside Advanced and then you don't need to expose the list.

Comment (TylerCarlson1) [inline src/AutoMapper/MapperConfiguration.cs]: Everything else goes through a foreach loop of items in this class.  Like profiles you can see the property but to add you call add function, so what's the difference?  It's exposed as an enumerable as to not add another way.  I don't see changing it as "that" big of a deal in exposing it or not.

Comment (lbargaoanu) [inline src/AutoMapper/MapperConfiguration.cs]: It's encapsulated, that's the difference. An event does this for example.

**Comment by TylerCarlson1** [comment]
Is this good @jbogard or is the before seal action implementation blocking it?

**Comment by jbogard** [comment]
I might do a wee bit of refactoring here but I don't want to block this one.

**Pull request merged**